hal/watchdog_api.h

```diff
@@ -0,0 +1,167 @@
+   * defined by window_ms. This value is false by default.
+   */
+  bool enable_window;
+  /**
+   * Specifies the time window for the watchdog window in milliseconds. If the
+   * watchdog is configured to run in windowed mode kicking the watchdog less
+   * than this many milliseconds after it has last been kicked will trigger a
+   * system reset. This value must be less than timeout_ms.
+   */
+  uint32_t window_ms;
+  /**
+   * Configures the watchdog timer to run while the core is in sleep mode. By
+   * default when the system is put into the sleep the watchdog timer is paused.
+   * Enabling this setting causes the timer to countdown during this time. This
+   * flag is disabled by default.
+   */
+  bool enable_sleep;
+} watchdog_config_t;
+
+
+typedef enum {
+  WATCHDOG_STATUS_OK,
+  WATCHDOG_STATUS_NOT_SUPPORTED,
+  WATCHDOG_STATUS_INVALID_ARGUMENT
+} watchdog_status_t;
+
+#ifdef __cplusplus
+    extern "C" {
+#endif
+
+/** Initialise and start a watchdog timer with the given configuration.
+ *
+ * If the watchdog timer is configured and started successfully this
+ * function will return WATCHDOG_STATUS_OK.
+ *
+ * If the enable_window is set but windowed mode is not supported by the
+ * platform the function will return WATCHDOG_STATUS_NOT_SUPPORTED.
+ *
+ * If the timeout specified is outside the range supported by the platform,
+ * or if the window period is greater than the timeout period it will return
+ * WATCHDOG_STATUS_INVALID_ARGUMENT.
+ *
+ * @param[in]  config   Configuration settings for the watchdog timer
+ *
+ * @return WATCHDOG_STATUS_OK if the watchdog is configured correctly and
+ *         has started. Otherwise a status indicating the fault.
+ */
+watchdog_status_t hal_watchdog_init(const watchdog_config_t *config);
+
+/** Refreshes the watchdog timer.
+ *
+ * This function should be called periodically before the watchdog times out.
+ * Otherwise, the system is reset.
+ *
+ * If using the windowed operation mode this function must be called within the
+ * configured timeout window. If the function is called before the window is
+ * reached the system is reset.
+ *
+ * If a watchdog is not currently running this function does nothing
+ */
+void hal_watchdog_kick(void);
+
+/** Stops the watchdog timer
+ *
+ * Calling this function will attempt to disable any currently running watchdog
+ * timers if supported by the current platform
+ *
+ * @return Returns WATCHDOG_STATUS_OK if the watchdog timer was succesfully
+ *         stopped, or if the timer was never started. Returns
+ *         WATCHDOG_STATUS_NOT_SUPPORTED if the watchdog cannot be disabled on
+ *         the current platform.
+ */
+watchdog_status_t hal_watchdog_stop(void);
+
+/** Get the watchdog timer refresh value
+ *
+ * This function returns the configured refresh timeout of the watchdog timer.
+ *
+ * @return Reload value for the watchdog timer in milliseconds.
+ */
+uint32_t hal_watchdog_get_reload_value(void);
```

Comment:
What if the windowed mode is used; wouldn't it be better to return the whole `watchdog_config_t` instead?

Reply:
I had considered that, it would make sense to provide a `get_current_config` function, though the config isn't currently stored anywhere, so it either needs to be saved somewhere in the API implementation or reformed from the watchdog registers

Reply:
I don't really see a good use case for retrieving the settings, that app needs to provide anyway. I think what Steve did with get is to map SDK funcionality to mbed.